snakemake/executors/flux.py

```diff
@@ -0,0 +1,266 @@
+
+    def cancel(self):
+        """
+        cancel execution, usually by way of control+c. Cleanup is done in
+        shutdown (deleting cached workdirs in Google Cloud Storage
+        """
+        for job in self.active_jobs:
+            if not job.flux_future.done():
+                flux.job.cancel(self.f, job.jobid)
+        self.shutdown()
+
+    def _set_job_resources(self, job):
+        """
+        Given a particular job, generate the resources that it needs,
+        including default regions and the virtual machine configuration
+        """
+        mem_mb = job.resources.get("mem_mb", 15360)
+
+        # Update default resources
+        self.default_resources = DefaultResources(
+            from_other=self.workflow.default_resources
+        )
+        self.default_resources.set_resource("mem_mb", mem_mb)
+
+    def get_snakefile(self):
+        assert os.path.exists(self.workflow.main_snakefile)
+        return self.workflow.main_snakefile.replace(self.workdir, "").strip(os.sep)
+
+    def _get_jobname(self, job):
+        # Use a dummy job name (human readable and also namespaced)
+        return "snakejob-%s-%s-%s" % (self.run_namespace, job.name, job.jobid)
+
+    def _generate_environment(self):
+        """
+        Loop through envvars (keys to host environment) and add
+        any that are requested for the runtime environment
+        """
+        envvars = dict(os.environ)
+        for key in self.envvars:
+            try:
+                envvars[key] = os.environ[key]
+            except Exception:
+                continue
+        return envvars
+
+    def format_job_exec(self, job):
+        """
+        We don't want a python -m to run.
+        """
+        prefix = self.get_job_exec_prefix(job)
+        if prefix:
+            prefix += " &&"
+        suffix = self.get_job_exec_suffix(job)
+        if suffix:
+            suffix = f"&& {suffix}"
+        return join_cli_args(
+            [
+                prefix,
+                self.get_envvar_declarations(),
+                "snakemake",
+                format_cli_arg("--snakefile", self.get_snakefile()),
+                self.get_job_args(job),
+                self.general_args,
+                suffix,
+            ]
+        )
+
+    def run(self, job, callback=None, submit_callback=None, error_callback=None):
+        """
+        Submit a job to flux.
+        """
+        super()._run(job)
+
+        from flux.job import JobspecV1
+
+        # This needs to be called after the line above
+        self._prepare_job_formatter()
+
+        # Prepare job resourcces
+        self._set_job_resources(job)
+
+        # The entire snakemake command to run, etc
+        command = self.format_job_exec(job)
+        nodes = job.resources.get("_nodes")
+        logger.debug(command)
+
+        # Generate the flux job
+        fluxjob = JobspecV1.from_command(
+            command=shlex.split(command),
+            num_nodes=nodes,
```

Comment:
Don't we need to pass additional resources to flux? mem, disk, etc.?

Reply:
Very likely! Here are the options:

https://github.com/flux-framework/flux-core/blob/0fa713c19b1f51b76d9f74063657b9fd54f8e60f/src/bindings/python/flux/job/Jobspec.py#L568-L571

Which should I try?

Reply:
I think we should let flux know about mem_mb, disk_mb, and runtime, see here: https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#standard-resources

Reply:
@johanneskoester - I checked with @grondo and out of those three, flux only supports runtime, and I've added this as a note to the docs (and that the default is 0 seconds, and should be provided in seconds).

Reply:
Our model supports all of the above, but our interface is currently limited.  Worth noting that will likely change.

Reply:
Maybe I can help here too. <3